From e79fb898dd7dd149da93c5218a71f24874c77282 Mon Sep 17 00:00:00 2001
From: Dinesh Karthikeyan <quic_dinek@quicinc.com>
Date: Tue, 30 May 2023 16:52:07 +0530
Subject: [PATCH] hostapd: Prioritize setting interface with DFS channel in MLD

Check if the interface has DFS channel in the MLD and start the CAC
for these interface first and then setup other radios.

Signed-off-by: Dinesh Karthikeyan <quic_dinek@quicinc.com>
---
 src/ap/hostapd.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 7959938..42ed4f3 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2680,6 +2680,38 @@ int hostapd_setup_interface_complete(struct hostapd_iface *iface, int err)
 	}
 #endif
 
+	for (i = 0; i < interfaces->count; i++) {
+		if (interfaces->iface[i]->need_to_start_in_sync &&
+		    interfaces->iface[i]->ready_to_start_in_sync) {
+			/* When DFS channel is present in MLD group, prioritize the order for
+			 * completing the  setup interface for that DFS vap first.
+			 */
+			if(!hostapd_is_dfs_required(interfaces->iface[i]))
+				continue;
+			hostapd_setup_interface_complete_sync(
+				interfaces->iface[i], 0);
+
+			/* Only once the interfaces are sync started */
+			interfaces->iface[i]->need_to_start_in_sync = 0;
+
+			/* During MLO, other link beaconing is dependent on this
+			 * iface state coming to ENABLED.
+			 *
+			 * This will be the case 5 GHz link is NOT the last config
+			 * provided in hostapd start command in order.
+			 * Example: 5GHz + 2 GHz
+			 */
+#ifdef CONFIG_IEEE80211BE
+			if ((interfaces->iface[i]->state == HAPD_IFACE_DFS) &&
+			     interfaces->iface[i]->bss[0]->mld) {
+				interfaces->iface[i]->need_to_start_in_sync = 1;
+				interfaces->iface[i]->ready_to_start_in_sync = 0;
+			}
+#endif
+		}
+	}
+
+
 	for (i = 0; i < interfaces->count; i++) {
 		if (interfaces->iface[i]->need_to_start_in_sync &&
 		    interfaces->iface[i]->ready_to_start_in_sync) {
-- 
2.17.1

